nexus/db-queries/src/db/datastore/db_metadata.rs

```diff
@@ -340,6 +662,252 @@ impl DataStore {
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        Ok(())
+    }
+
+    /// Initializes Nexus database access records from a blueprint using an
+    /// existing connection
+    ///
+    /// This function finds all Nexus zones in the given blueprint and creates
+    /// active database access records for them. Used during RSS rack setup.
+    ///
+    /// Returns an error if:
+    /// - Any db_metadata_nexus records already exist (should only be called
+    /// during initial setup)
+    pub async fn initialize_nexus_access_from_blueprint_on_connection(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        blueprint_id: Uuid,
+    ) -> Result<(), Error> {
+        use nexus_db_model::ZoneType;
+        use nexus_db_schema::schema::bp_omicron_zone::dsl as zone_dsl;
+        use nexus_db_schema::schema::db_metadata_nexus::dsl;
+
+        // Ensure no db_metadata_nexus records already exist
+        let any_records_exist =
+            Self::database_nexus_access_any_exist_on_connection(conn).await?;
+        if any_records_exist {
+            return Err(Error::internal_error(
+                "Cannot initialize Nexus access from blueprint: db_metadata_nexus records already exist. \
+                This function should only be called during initial rack setup.",
+            ));
+        }
+
+        // Query all Nexus zones from the blueprint
+        let nexus_zone_ids: Vec<Uuid> = zone_dsl::bp_omicron_zone
+            .filter(zone_dsl::blueprint_id.eq(blueprint_id))
+            .filter(zone_dsl::zone_type.eq(ZoneType::Nexus))
+            .select(zone_dsl::id)
+            .load_async(conn)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        // Create db_metadata_nexus records for all Nexus zones
+        let new_nexuses: Vec<DbMetadataNexus> = nexus_zone_ids
+            .iter()
+            .map(|&nexus_id| {
+                DbMetadataNexus::new(
+                    OmicronZoneUuid::from_untyped_uuid(nexus_id),
+                    DbMetadataNexusState::Active,
+                )
+            })
+            .collect();
+
+        diesel::insert_into(dsl::db_metadata_nexus)
+            .values(new_nexuses)
+            .execute_async(conn)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        Ok(())
+    }
+
+    // Implementation function for attempt_handoff that runs within a transaction
+    //
+    // This function must be executed from a tranaction context to be safe.
+    async fn attempt_handoff_impl(
+        conn: async_bb8_diesel::Connection<DbConnection>,
+        nexus_id: OmicronZoneUuid,
+        err: OptionalError<HandoffError>,
+    ) -> Result<(), diesel::result::Error> {
+        use nexus_db_schema::schema::db_metadata_nexus::dsl;
+
+        // Before proceeding, all records must be in the "inactive" or "not_yet" states.
```

Comment:
```suggestion
        // Before proceeding, all records must be in the "inactive" or "not_yet" states.
        // Again, there's no time-of-check-to-time-of-use race here because we're in a
        // transaction. 
```

Reply:
Updated